File: python/frogpath.py
def count_paths(begin_point, end_point):
    return count_all([(begin_point, '')], end_point)


def count_all(begin_points, end_point):
    ended_num = 0
    next_points = []
    for point, last_step in begin_points:
        if point == end_point:
            ended_num += 1
        else:
            next_points += move_step(point, last_step, end_point)
    if next_points:
        return ended_num + count_all(next_points, end_point)
    return ended_num


def move_step(begin_point, last_step, end_point):
    next_points = []
    for direction in ['u', 'r', 'd']:
        if not (direction == 'r' and last_step == 'u' or direction == 'u' and last_step == 'r'):
            new_point = move_direction(begin_point, direction)
            if new_point[0] <= end_point[0] and new_point[1] <= end_point[1]:
                next_points.append((new_point, direction))
    return next_points


def move_direction(begin_point, direction):
    x, y = begin_point
    if direction == 'u':
        return x, y + 1
    if direction == 'r':
        return x + 1, y
    return x + 1, y + 1

File: python/frogpath.ml
type direction = Right | Up | Diagnal | None

let move_direction (x, y) = function
  | Right -> x + 1, y
  | Up -> x, y + 1
  | Diagnal -> x + 1, y + 1
  | _ -> failwith "invalid direction!"

let move_step begin_point last_step (end_x, end_y) cur_points =
  List.fold_left (fun points direction ->
      match direction, last_step with
      | Right, Up
      | Up, Right -> points
      | _ ->
        let (x, y) as new_point = move_direction begin_point direction in
        if x <= end_x && y <= end_y then
          (new_point, direction)::points
        else points)
    cur_points
    [Right; Up; Diagnal]

let rec count_all begin_points end_point num =
  let next_points, ended_num = List.fold_left
      (fun (points, num) (point, last_direction) ->
         if point = end_point then (points, num + 1)
         else
           move_step point last_direction end_point points, num
      )
      ([], num)
      begin_points
  in match next_points with
  | [] -> ended_num
  | _ -> count_all next_points end_point ended_num

let count_paths begin_point end_point =
  count_all [(begin_point, None)] end_point 0

let () =
  let x = count_paths (0,0) (15, 15) in
  print_int x
